### PR TITLE
Add specifications for sine, cosine, and tanh operators

### LINKED ISSUE #85: Add the specifications of sine/cosine/tanh operators

### PR DESCRIPTION
**Pull request by ghpvnist**
fixes #85

### CODE REVIEW

File: docs/spec_draft.md [hunk 1]
```diff
@@ -223,6 +224,48 @@ Produces a `result` tensor from a constant `value`.
...
+
+| Name | Type |
+|-|-|
+| `value` | Constant n-dimensional data of floating-point, or complex element types |
+
+### Results
+
+| Name | Type |
+|-|-|
+| `result` | `tensor` of floating-point, or complex element types |
```
Comment (sdasgup3) [inline docs/spec_draft.md:251]: remove the "," when having just two items? Here and other similar places

File: docs/spec_draft.md [hunk 2]
```diff
@@ -223,6 +224,48 @@ Produces a `result` tensor from a constant `value`.
...
 
+## stablehlo.cosine
+
+`stablehlo.cosine(operand) -> result`
+
+### Semantics
+
+Performs element-wise cosine operation on `operand` tensor and produces a `result` tensor.
+
+### Attributes
```
Comment (sdasgup3) [inline docs/spec_draft.md:235]: We do not need an Attribute for this op.

File: docs/spec_draft.md [hunk 3]
```diff
@@ -223,6 +224,48 @@ Produces a `result` tensor from a constant `value`.
...
+
+| Name | Type |
+|-|-|
+| `value` | `tensor` of floating-point or complex element types |
+
+### Operands
+
+| Name | Type |
+|-|-|
+| `value` | Constant n-dimensional data of floating-point, or complex element types |
```
Comment (sdasgup3) [inline docs/spec_draft.md:245]: operand instead of value?

Comment (sdasgup3) [inline docs/spec_draft.md:245]: We should use tensor which is already defined as n dimensional data. Something like

`tensor of floating-point or complex element types`

Comment (sdasgup3) [reply docs/spec_draft.md:245]: here and below

Comment (sdasgup3) [reply docs/spec_draft.md:245]: here and below

File: docs/spec_draft.md (hunk 2 above)
Comment (sdasgup3) [reply docs/spec_draft.md:235]: here and below

**Review by sdasgup3 — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md [hunk 4]
```diff
@@ -223,6 +224,48 @@ Produces a `result` tensor from a constant `value`.
 // %3: [(0.0, 1.0), (2.0, 3.0)]
 ```
 
+## stablehlo.cosine
+
+`stablehlo.cosine(operand) -> result`
+
+### Semantics
+
+Performs element-wise cosine operation on `operand` tensor and produces a `result` tensor.
```
Comment (burmako) [inline docs/spec_draft.md:233]: What does it mean to perform a cosine operation? What kind of precision with respect to an ideal mathematical result are we talking about? What are the corner cases? (Same questions for other ops in this PR and for sqrt/rsqrt in the neighboring PR).

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 1 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Fixed

File: docs/spec_draft.md (hunk 2 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Fixed

File: docs/spec_draft.md (hunk 3 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Fixed

Comment (ghpvnist) [reply docs/spec_draft.md]: Fixed

File: docs/spec_draft.md [hunk 5]
```diff
@@ -405,6 +448,80 @@ operation.
...
+[Back to Ops](#index-of-documented-ops)
+
+## stablehlo.tanh
+
+`stablehlo.tanh(operand) -> result`
+
+### Semantics
+
+Performs element-wise tanh operation on `operand` tensor and produces a `result` tensor.
+$$tanh(x)=\frac{e^{x}-e^{-x}}{e^{x}+e^{-x}}$$
```
Comment (burmako) [inline docs/spec_draft.md:496]: What should be our policy for providing explanations on what some of these operations do? How about we consistently refer to IEEE-754 if they are part of it, or express them in terms of IEEE-754 operations if they aren't?

E.g. `tanh` is called out in IEEE-754, so we could just say "Performs element-wise tanh operation on `operand` tensor and produces a `result` tensor, according to the IEEE-754 specification"?

Comment (burmako) [inline docs/spec_draft.md:496]: On a separate note, as discussed offline, let's also call out numeric precision for floating-point ops. Initially, that could be as simple as "Numeric precision is implementation-defined".

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 4 above)
Comment (ghpvnist) [reply docs/spec_draft.md:247]: As mentioned in later comments, we will initially state, "Numeric precision is implementation-defined."

File: docs/spec_draft.md [hunk 6]
```diff
@@ -405,6 +450,84 @@ operation.
...
+## stablehlo.tanh
+
+`stablehlo.tanh(operand) -> result`
+
+### Semantics
+
+Performs element-wise tanh operation on `operand` tensor and produces a `result`
+tensor, according to the IEEE-754 specification. Numeric precision is
+implementation-defined.
+$$tanh(x)=\frac{e^{x}-e^{-x}}{e^{x}+e^{-x}}$$
```
Comment (burmako) [inline docs/spec_draft.md:502]: Let's drop the formula, given that we're referenced the specification?

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 6 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Done

File: docs/spec_draft.md (hunk 5 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Done

**Review by burmako — APPROVED** [review]

File: docs/spec_draft.md [hunk 7]
```diff
@@ -277,6 +280,46 @@ Produces a `result` tensor from a constant `value`.
...
 [Back to Ops](#index-of-documented-ops)
 
+## stablehlo.cosine
+
+`stablehlo.cosine(operand) -> result`
+
+### Semantics
+
+Performs element-wise cosine operation on `operand` tensor and produces a
+`result` tensor, according to the IEEE-754 specification. Numeric precision is
```
Comment (sdasgup3) [inline docs/spec_draft.md:290]: `implementing to the `cos` operation from the IEEE-754 specification.`
Same for other ops below.

File: docs/spec_draft.md [hunk 8]
```diff
@@ -277,6 +280,46 @@ Produces a `result` tensor from a constant `value`.
...
+implementation-defined.
+
+### Operands
+
+| Name | Type |
+|-|-|
+| `operand` | tensor of floating-point or complex element types |
```
Comment (sdasgup3) [inline docs/spec_draft.md:297]: Can we say something for the complex numbers? We can check with HLO?

For example for sine,
```
sine(a-+ib) = sin(a)*cosh(b) +i cos(a)*sinh(b)
```
Same for other ops.

https://proofwiki.org/wiki/Sine_of_Complex_Number/Proof_1

Comment (sdasgup3) [reply docs/spec_draft.md:297]: Ignore the previous comments. 
From the fact that this is a standard textbook information,  let us just skip mentioning it here in the spec. Same for ops below

File: docs/spec_draft.md (hunk 7 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: Done.

**Review by sdasgup3 — APPROVED** [review]

**Pull request merged**